rfcs/20221010-add-ocl-ooo-queue/README.md

```diff
@@ -0,0 +1,125 @@
+
+### Out-of-Order Stream
+oneDNN stream abstraction that encapsulates an OpenCL queue, in the case of OpenCL runtime,
+supports flags that controls stream behavior. The flags include `in_order` and `out_of_order`
+values that can be used to create an out-of-order stream. Currently, oneDNN returns a status
+`unimplemented` when a user tries to create an out-of-order stream for OpenCL runtime
+therefore no new API is required. The out-of-order stream will be made available for creation
+in the case of OpenCL runtime.
+
+### OpenCL Interop API
+The API is an extension of the generic execute API and has two additional responsibilities:
+1. Take a vector of events for dependencies
+2. Return an event for the primitive that was submitted for execution
+
+#### C API
+
+```c
+// dnnl_ocl.h
+
+/// Executes computations specified by the primitive in a specified stream and
+/// returns an OpenCL event.
+///
+/// @param primitive Primitive to execute.
+/// @param stream Stream to use.
+/// @param nargs Number of arguments.
+/// @param args Array of arguments. Each argument is an
+///     <index, #dnnl_memory_t> pair. The index is one of the `DNNL_ARG_*`
+///     values such as `DNNL_ARG_SRC`. Unless runtime shapes are used (see
+///     #DNNL_RUNTIME_DIM_VAL), the memory object must have the same memory
+///     descriptor as that returned by
+///     #dnnl_primitive_desc_query_md(#dnnl_query_exec_arg_md, index).
+/// @param deps A pointer to std::vector<cl_event> that contains
+///     dependencies.
+/// @param return_event Output event. The library doesn't own the event.
+///     It is the user's responsibility to manage lifetime of the event.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_ocl_interop_primitive_execute(
+        const_dnnl_primitive_t primitive, dnnl_stream_t stream, int nargs,
+        const dnnl_exec_arg_t *args, const cl_event *deps, int ndeps,
+        cl_event *return_event);
+```
+
+#### C++ API
+```cpp
+// dnnl_ocl.hpp
+
+/// Executes computations specified by the primitive in a specified stream and
+/// returns an OpenCL event.
+///
+/// Arguments are passed via an arguments map containing
+/// <index, memory object> pairs. The index must be one of the `DNNL_ARG_*`
+/// values such as `DNNL_ARG_SRC`, and the memory must have a memory descriptor
+/// matching the one returned by
+/// #dnnl::primitive_desc::query_md(#query::exec_arg_md, index) unless using
+/// dynamic shapes (see #DNNL_RUNTIME_DIM_VAL).
+///
+/// @param aprimitive Primitive to execute.
+/// @param astream Stream object. The stream must belong to the same engine
+///     as the primitive.
+/// @param args Arguments map.
+/// @param deps Optional vector with `cl::sycl::event` dependencies.
+///
+/// @returns Output event. The library doesn't own the event. It is the user's
+///     responsibility to manage lifetime of the event.
+inline cl_event execute(const dnnl::primitive &aprimitive,
```

Comment:
For DG2 sync via barriers seems to be more efficient than events. We can do the barrier-based sync on OV side, but seems like this API will always trigger creating of output event in onednn. Not sure that it will really impact the performance, but it'd be great to check this barriers story.

Reply:
Ok, in other words it would be better if we allowed to use the generic API with an out-of-order queue because such API doesn't generate any events and it would be the user's responsibility to synchronize the call, supposedly with a barrier. The in-order queue will also be support, of course.

The interop API would support both queue kinds too.